Overhaul Array documentation

Mickeon opened the pull request:
Closes https://github.com/godotengine/godot-docs/issues/6964
Closes https://github.com/godotengine/godot-docs/issues/8695
Closes https://github.com/godotengine/godot-docs/issues/9005
Closes https://github.com/godotengine/godot/issues/79066
Closes https://github.com/godotengine/godot/issues/89276 (properly)
Closes https://github.com/godotengine/godot-docs/issues/9067

Continuation of a series of PRs that go on to touch major, yet majorly unrevised class references. This includes #67072, #67100, #67208, #67718,... #67880 and #68560... and #68649... And I think I'll end up putting all of these in a more organised list at some point.

## General


- Made the wording match how other classes are documented a **lot** more closely;
- Made use of _`[param]`_ and several other strong references more often;
- Changed words where it felt necessary, where they could be mistaken for another concept of Godot's API, where technical terms are more appropriate. For example:
	- Use "element" consistently.
	- Use "index" consistently.
		- This is not fully possible yet because some parameters are named "position" and it would be confusing.
- Modified examples extensively, with the main goal to improve readability and variety:
	- Removed quotes around results when they could be misunderstood as strings (`# Prints "5"` -> `# Prints 5`);
	- Improved consistency all around;
	- Avoided big walls of `print()`;
	- Added extra padding before comments.
	- Avoid overly long lines.
- Added slightly more detail to the leading description;
- Moved the entire _"Arrays can be concatenated using the `+` operator"_ thing out of the leading description.

-----------------------------
## Constructors

**`Array(from)`**
- Updated description to be correct.
	- This is temporary. It's not something desired, and should either be modified to work as intended (copy), or removed.See https://github.com/godotengine/godot/pull/69117

---------------------
## Methods

### **`bsearch`**, **`bsearch_custom`**
- Simplified wording considerably.
- Linked to Binary Search as a bit of an ending note. This is to better understand why the method is named this way, but is not mandatory to know to make use of it.
- Added `bsearch` example;

### **`all`**, **`any`**, **`map`**, **`reduce`**
- Modified Examples:
- Show the Callable's definition first.
- Simplified methods with more familiar terms (`negate` -> `double`)

### **`clear`**
- _"Clears the array"_
	- Cool, thanks!

### **`contains`**
- Added note about how the check is done by exact Variant type _(e.g. `7` is not equal to `7.0`)_. 

### **`duplicate`**
- Rewrote shallow/deep explanation.

### **`erase`**, **`remove_at`**, **`insert`**, **`pop_at`**, **`pop_front`**
- _"This method acts in-place and doesn't return a value."_
Patronising...
- Added "Finds and removes" in erase. This slightly extra verbosity implies that the element's index is not known in advance, potentially being slower than removing by index.
- Simplified the _"On large arrays, this method will be slower if the..."_ note considerably.

### **`fill`**
- Modified example to be more "show", less "tell".
- Rewrote shallow/deep explanation.

### **`find`**, **`rfind`**
- Modified description entirely. These are based off https://github.com/godotengine/godot/pull/68838.

### **`hash`**
- Modified note entirely, based off https://github.com/godotengine/godot/pull/68838.

### **`is_typed`**
- Added static typing example.

### **`set_typed`**
- Expanded description of individual parameters (using `[br]`).

### **`shuffle`*
-  _"Call `randomize` to ensure that a new seed will be used each time if you want non-reproducible shuffling."_
In Godot 4, the seed is randomly generated once at start-up, so this is no longer necessary, but the "opposite" is.

### **`slice`*
- Modified wording considerably.
	- This method is such a bother to explain. It does something trivially simple, but the complication is negative parameters being perfectly allowed for some reason.
- Removed inline examples. They only made the whole thing more difficult to read.
- Added some examples.

###**`sort`**
- Add more detail to how sorting is performed
- Move String natural order sorting example out of this description.

### **`sort_custom`**
- ~Modified example to use nested Dictionaries over nested Arrays, to potentially be less confusing while keeping it brief.~
    - Reverted until a nicer solution is found, if possible.
- Moved String natural order sorting example here, previously in `sort` description.


------------------------
## Operators

### **`+`**
- Moved concrete concatenation example here, previously present in the leading description.
- Explained the performance hit from array concatenation more to the point.

### **`+`**, **`>`**, **`>=`**, **`<`**, **`<=`**
- Attempted to simplify the explanation considerably.
	- _**OH BOTHER THIS WAS SUCH A PAIN**_...

### **`[]`**
- Polished description (based off https://github.com/godotengine/godot/pull/68838).

-----------------------------------------
Feedback is very, very welcome.

Mickeon committed:
Overhaul Array Documentation

Mickeon commented on doc/classes/Array.xml:
```diff
@@ -183,36 +174,46 @@
-				[b]Note:[/b] Calling this function is not the same as writing [code]array[-1][/code]. If the array is empty, accessing by index will pause project execution when running from the editor.
+				?? Returns the last element of the array. Generates an error and returns [code]null[/code] if the array is empty.
+				?? Returns the last element of the array. If the array is empty, generates an error and returns [code]null[/code], without stopping the project execution (unlike with the [code][][/code] operator).
+				?? [b]Note:[/b] This method is not the same as [code]array[-1][/code].
```
Unsure of how to say this, here. I do not want to lose  _"Calling this function is not the same as writing [code]array[-1][/code]"_, but it also feels like repetitive information at the same time.

Mickeon commented on doc/classes/Array.xml:
```diff
@@ -183,36 +174,46 @@
-				[b]Note:[/b] Calling [method bsearch] on an unsorted array results in unexpected behavior.
+				Returns the index of [param value] in the sorted array. If it cannot be found, returns where [param value] should be inserted to keep the array sorted. The algorithm used is [url=https://en.wikipedia.org/wiki/Binary_search_algorithm]binary search[/url].
+				?? If [param before] is [code]false[/code], the returned index comes after all existing entries of the [param value] in the array.
+				?? If [param before] is [code]true[/code] (as by default), the returned index comes before all elements that are equal to [param value] in the array.
```
It's a bit weird to say "before is false", and then say "after". May be just me, I just am not sure.

Mickeon replied:
Only minor problem is, I actively avoided calling them entries on purpose. It's not an incorrect term, it was just a matter of consistency in making sure it's always "element".

Mickeon commented on doc/classes/Array.xml:
```diff
@@ -297,88 +303,83 @@
-				Returns the first element of the array. Prints an error and returns [code]null[/code] if the array is empty.
-				[b]Note:[/b] Calling this function is not the same as writing [code]array[0][/code]. If the array is empty, accessing by index will pause project execution when running from the editor.
+				?? Returns the first element of the array. If the array is empty, generates an error and returns [code]null[/code], similarly to [], without stopping the project execution.
+				?? Returns the first element of the array. If the array is empty, generates an error and returns [code]null[/code] without stopping the project execution (unlike with the [code][][/code] operator).
```
Same as above.
Unsure of how to say this, here. I do not want to lose "Calling this function is not the same as writing [code]array[0][/code]", but it also feels like repetitive information at the same time.

souplamp replied:
If you had to pick from these options, the second line is a lot more clear at demonstrating the point. 
The use of `unlike` clarifies how the function will do something different than regularly accessing the array.
The first line uses `similarly`, which implies neither stop project execution, which gives the wrong impression.

Mickeon commented on doc/classes/Array.xml:
```diff
@@ -297,88 +303,83 @@
 			<description>
-				Returns [code]true[/code] if the array contains the given value.
+				?? Returns [code]true[/code] if the array contains the given [param value].
+				?? Returns [code]true[/code] if any of the array's elements is the given [param value].
```
The original sentence is fine to me and maybe most users, but it doesn't really explain what "contains" is. The alternative does that, but is ever so slightly more verbose. So, I'm not sure.

souplamp replied:
I would say that 'contains' is clear as is. Just from the name of the function, users should pick up on the concept.

MewPurPur replied:
No need to be verbose IMO

Mickeon replied:
For me there's a slight need to, but mostly because I want it to translate appropriately, too. On the other hand, the concept of "to have" or "contain" something is pretty dang global.

Mickeon commented on doc/classes/Array.xml:
```diff
@@ -297,88 +303,83 @@
+				?? [b]Note:[/b] The index of every element after [param position] is shifted forward. This means that this method is slower the closer [param position] is to, which may be especially noticeable on very large arrays.
+				?? [b]Note:[/b] This method is slower the closer [param position] is to the beginning of the array. This is because every element's index after [param position] is shifted forward, which may be especially slow on very large arrays.
+				?? [b]Note:[/b] This method shifts every element's index after [param position] forward, which may have a noticeable performance cost, especially on larger arrays.
+				?? [b]Note:[/b] Every element's index after [param position] needs to be shifted forward, which may have a noticeable performance cost, especially on larger arrays.
```
Yes. 5 whole variations for this sentence. The last is the one I used elsewhere too because it is the simplest to understand in concept, but I'm not sure if that would be losing too much information.

Mickeon commented on doc/classes/Array.xml:
```diff
@@ -496,37 +509,39 @@
 			<description>
-				Resizes the array to contain a different number of elements. If the array size is smaller, elements are cleared, if bigger, new elements are [code]null[/code].
+				?? Resizes the array to contain a different number of elements. If [param size] is smaller than the array's current size, the elements at the end are removed. Otherwise, the new elements are [code]null[/code].
+				?? Sets the array's number of elements to [param size]. If [param size] is smaller than the array's current size, the elements at the end are removed. If [param size] is greater, new [code]null[/code] elements are added.
```
The latter revision is more correct and less repetitive, but the former revision does not sound too bad, either.

Mickeon commented on doc/classes/Array.xml:
```diff
@@ -666,21 +712,27 @@
+				?? Compares each element of the array against the element of [param right], in order. Returns [code][/code]
+				?? Performs a "greater than" comparison between the elements of both arrays, starting from index [code]0[/code] and ending on the last index in common between both arrays. Returns [code]true[/code] if an element is greater than [right] and false. Otherwise, if equal, continues to the next index.
+				?? Compares the elements of both arrays in order, starting from index [code]0[/code] and ending on the last index in common between both arrays. For each pair of elements, returns [code]true[/code] if this array's element is greater than [param right]'s, [code]false[/code] if this element is lower. Otherwise, continues to the next pair.
+				If all searched elements are equal, returns [code]true[/code] if this array's size is greater than [param right]'s, otherwise returns [code]false[/code].
```
Yes, this entire explanation is a bloody beast.

Mickeon commented on doc/classes/Array.xml:
```diff
@@ -138,18 +127,19 @@
 			<description>
 				Calls the provided [Callable] on each element in the array and returns [code]true[/code] if the [Callable] returns [code]true[/code] for [i]all[/i] elements in the array. If the [Callable] returns [code]false[/code] for one array element or more, this method returns [code]false[/code].
-				The callable's method should take one [Variant] parameter (the current array element) and return a boolean value.
+				The [param method] should take one [Variant] parameter (the current array element) and return a boolean value.
```
All of the methods that accept Callables have this problem where their wording to describe the Callable parameter is inconsistent, but I am not sure exactly what the ideal way to do it would be.

It doesn't help that currently some of these parameters are called "method" but I digress for now. That can be addressed another time.

Mickeon replied:
In reality it would be a good idea to move what the Callable should *be* to the second paragraph. That is the main inconsistency issue.

adamscott commented on doc/classes/Array.xml:
```diff
@@ -138,18 +127,19 @@
 				Calls the provided [Callable] on each element in the array and returns [code]true[/code] if the [Callable] returns [code]true[/code] for [i]all[/i] elements in the array. If the [Callable] returns [code]false[/code] for one array element or more, this method returns [code]false[/code].
-				The callable's method should take one [Variant] parameter (the current array element) and return a boolean value.
+				The [param method] should take one [Variant] parameter (the current array element) and return a boolean value.
 				[codeblock]
```
Shouldn't it be `[gdscript]` instead of `[codeblock]` here and in other GDScript examples?

Mickeon replied:
Many of Array's methods are not available in C#'s Godot.Collections.Array. At least, not in a way I'm sure can easily be translated.

adamscott replied:
But the codeblock itself should highlight gdscript code. even if this is only showing GDScript.

Mickeon replied:
So the C# code example should exist, but either be "This does not exist?" or provide an alternative? Not sure about it.

Maybe @YuriSizov can give more of an opinion. There was a similar argument to be had in https://github.com/godotengine/godot/pull/67880#discussion_r1030505866:
> Also also, it's weird to say "In GDScript" and give examples in both GDScript and C#, with the latter using a different syntax 🙂

adamscott replied:
Nevermind, I thought that it was possible to use `[gdscript]` in a standalone way, without `[codeblocks]`, but it's not possible.

raulsntos replied:
Even when the example is provided only for GDScript, I think it's still better to use `[codeblocks][gdscript]` because then it shows the tab that makes it more clear that the code is in GDScript.

About translating examples to C#, the table in the [_C# Collections_](https://docs.godotengine.org/en/stable/tutorials/scripting/c_sharp/c_sharp_collections.html) documentation page should help with that, but let me know if there are any code examples that you are not sure how to translate.

In this example, you can use the Linq [`All`](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.all) method which is very similar to the GDScript version.

Mickeon replied:
Could you provide a similar example to the GDScript one used, for GDScript's arrays?

raulsntos replied:
```csharp
private static bool GreaterThan5(int number)
{
	return number > 5;
}

public override void _Ready()
{
	GD.Print(new Godot.Collections.Array<int> { 6, 10, 6 }.All(GreaterThan5)); // Prints true (3/3 elements evaluate to true).
	GD.Print(new Godot.Collections.Array<int> { 4, 10, 4 }.All(GreaterThan5)); // Prints false (1/3 elements evaluate to true).
	GD.Print(new Godot.Collections.Array<int> { 4, 4, 4 }.All(GreaterThan5)); // Prints false (0/3 elements evaluate to true).
	GD.Print(new Godot.Collections.Array<int> { }.All(GreaterThan5)); // Prints true (0/0 elements evaluate to true).

	// Same as the first line above, but using a lambda function.
	GD.Print(new Godot.Collections.Array<int> { 6, 10, 6 }.All(element => element > 5)); // Prints true
}
```

MewPurPur commented on doc/classes/Array.xml:
```diff
@@ -297,88 +303,83 @@
 			<return type="StringName" />
 			<description>
-				Returns a class name of a typed [Array] of type [constant TYPE_OBJECT].
+				Returns the base class name of the typed [Array] of type [constant TYPE_OBJECT]. See also [method set_typed].
```
Reference to Array seems unnecessary here.

MewPurPur commented on doc/classes/Array.xml:
```diff
@@ -297,88 +303,83 @@
+				In GDScript, this is equivalent to the [code]in[/code] operator:
+				[codeblock]
+				if 4 in [2, 4, 6, 8]:
+				    print("There it is!")
```
A comment should make clear that this line will run.

MewPurPur commented on doc/classes/Array.xml:
```diff
@@ -390,73 +391,82 @@
+				Returns [code]true[/code] if the array is typed. Typed arrays can only contain elements of a specific type, as defined by the [method set_typed] method, or typed array constructor. The methods of a typed array are still expected to return a generic [Variant].
+				In GDScript, it is possible to define a typed array with static typing:
+				[codeblock]
+				var numbers : Array[float] = [0.2, 4.2, -2.0]
```
GDScript's style guide asks that there be no spaces before colons.

MewPurPur commented on doc/classes/Array.xml:
```diff
@@ -535,14 +550,23 @@
+				In GDScript, this method is usually not necessary, as it is possible to create a typed array through static typing:
+				[codeblock]
+				var numbers : Array[float] = [0.2, 4.2, -2.0]
+				var children : Array[Node] = get_children()
```
Same remark, colons should have no space before.

MewPurPur commented on doc/classes/Array.xml:
```diff
@@ -558,71 +582,74 @@
+				    var my_items = [{"key": 5}, {"key": 9}, {"key": 4}]
 				    my_items.sort_custom(sort_ascending)
-				    print(my_items) # Prints [[4, Tomato], [5, Potato], [9, Rice]].
+				    print(my_items) # Prints [{"key": 4}, {"key": 5}, {"key": 9}]
```
Why not just show an implementation of quicksort in GDScript? /joke

You're comparing values here, not keys, I believe.

Mickeon replied:
Yeah, that's not great to hear. Indeed. The comparison is done with every value of the "key" entry in each Dictionary. May be worth reverting this back to an Array. The most complete example of custom sorting would be pretty long and verbose, and I'd avoid that, but it'd be nice to find a good balance.

MewPurPur replied:
Oh I get it now. Yeah - it's confusing IMO.

Mickeon commented on doc/classes/Array.xml:
```diff
@@ -558,71 +581,74 @@
+				If either [param begin] or [param end] are negative, their value is relative to the end of the array.
+				If [param step] is negative, this method iterates through the array in reverse, returning a slice ordered backwards. For this to work, [param begin] must be higher than [param end].
+				?? If [param deep] is [code]true[/code], each element is copied by value rather than by reference. This means that nested Dictionaries and other Arrays are duplicated (see also [method duplicate]).
+				?? If [param deep] is [code]true[/code], all nested [Array] and [Dictionary] elements in the slice are duplicated from the original, recursively. See also [method duplicate]).
```
Hmmmm....

Mickeon commented on doc/classes/Array.xml:
```diff
@@ -631,28 +657,47 @@
+				print(array1 + array2) # Prints ["One", 2, 3, "Four"]
+				[/gdscript]
+				[csharp]
+				// Note that concatenation is not possible with C# arrays.
```
The original description states that it is not possible to concatenate C#'s native array type, but it is possible to concatenate Godot.Collections.Array.

Mickeon commented:
Rebased.

raulsntos commented on doc/classes/Array.xml:
```diff
@@ -250,53 +259,57 @@
 			<description>
-				Returns a copy of the array.
-				If [param deep] is [code]true[/code], a deep copy is performed: all nested arrays and dictionaries are duplicated and will not be shared with the original array. If [code]false[/code], a shallow copy is made and references to the original nested arrays and dictionaries are kept, so that modifying a sub-array or dictionary in the copy will also impact those referenced in the source array. Note ...
+				Returns a new copy of the array.
+				By default, a [b]shallow[/b] copy is returned: all nested [Array] and [Dictionary] elements are shared with the original array. Modifying them in one array will also affect them in the other.[br]If [param deep] is [code]true[/code], a [b]deep[/b] copy is returned: all nested arrays and dictionaries are also duplicated (recursively).
```
There was some feedback about the description of the `duplicate` method in https://github.com/godotengine/godot/pull/71786#discussion_r1085737667.

@RedworkDE Do you think the description is better now?

RedworkDE replied:
I'd say it better, but the (incorrect) note about objects got lost, instead of corrected.
Also we may want to elaborate on the meaning of shared, unless that is clarified elsewhere already.
Additionally the behavior when the original array contains shared elements may be unexpected.
I left suggestions on how these things could be worded but that probably needs a second (and third pass)
One last thing the recursion depth of deep copies is also limited, but I am not sure if and how this should be documented here.
```suggestion
				Returns a new copy of the array.
				By default, a [b]shallow[/b] copy is returned: all nested [Array] and [Dictionary] elements are shared with the original array. Modifying them in one array will also affect them in the other.[br]If [param deep] is [code]true[/code], a [b]deep[/b] copy is returned: all nested arrays and dictionaries are also duplicated (recursively).
				[b]Note:[/b] [Object]-derived elements are always shared with the original array regardless of the [param deep] setting.
				[b]Note:[/b] An element is shared iff a modification to one version of it is also visible to the other.
				[b]Note:[/b] If the original contained shared elements, after a deep copy, these elements will no longer be shared and instead be distinct copies.
```

Mickeon replied:
I want to note that it's worth being really careful with words, and how much to really expose to the documentation. In my past contributions to the docs I have always avoided to express way too much detail. It is quite overwhelming to any user, and they make the subject look more complex than it truly is. Oftentimes these concepts are better explained somewhere else, and merely reminded of where they matter.

Mickeon commented:
Addressed spell check fails.

Mickeon commented:
Rebased.

I need to note a noteworthy change during the rebase.

An example was present in `max()`, explaining how to use `reduce()` for a custom comparator. I shifted this example within `reduce()` 's description, and modified it a bit to be slightly more intuitive. This attempt is to not lose possibly valuable information that was added because of a past report, _although I would've personally changed the example outright_.

AThousandShips commented on doc/classes/Array.xml:
```diff
@@ -648,28 +673,44 @@
 			<param index="0" name="right" type="Array" />
 			<description>
-				Performs a comparison for each index between the left operand [Array] and the [param right] [Array], considering the highest common index of both arrays for this comparison: Returns [code]true[/code] on the first occurrence of an element that is less, or [code]false[/code] if the element is greater. Note that depending on the type of data stored, this function may be recursive. If all element ...
+				Compares the elements of both arrays in order, starting from index [code]0[/code] and ending on the last index in common between both arrays. For each pair of elements, returns [code]true[/code] if this array's element is lower than [param right]'s, [code]false[/code] if this element is greater. Otherwise, continues to the next pair.
```
```suggestion
				Compares the elements of both arrays in order, starting from index [code]0[/code] and ending on the last index in common between both arrays. For each pair of elements, returns [code]true[/code] if this array's element is less than [param right]'s, [code]false[/code] if this element is greater. Otherwise, continues to the next pair.
```
Same elsewhere

Mickeon replied:
Replaced every instance of lower with less.

Mickeon commented:
Rebased after https://github.com/godotengine/godot/pull/80936 and a few more I believe.
Of note is that in this PR, most of the _"This method acts in-place."_ notes do not actually exist because the description itself _should_ make it clearer. Please take note of this and do express when it should be re-added.

Mickeon commented:
Updated PR to add a C# example as suggested by raulsntos.

Mickeon commented:
Rebased to force the Linux test to start, hopefully without bugging out.

Mickeon commented:
Rebased after:
- https://github.com/godotengine/godot/pull/84666

Mickeon commented:
Rebased.

donn-xx commented:
Under **bsearch**, can anyone write something to explain what this means:

    "(or the insertion index that maintains sorting order, if the value is not yet present in the array)"

The sense of it does not leap from the screen. What is an 'insertion index'? What is 'sorting order'? Why would a value not be in the array yet?

😄

Mickeon commented:
Please click on **Files changed** to see what this PR does and to make accurate comparisons.
![image](https://github.com/godotengine/godot/assets/66727710/be0f608b-5a9e-4fdf-aadc-c554cacf1702)

In this PR, I have already modified the description a loooong while ago, (in an attempt) to be more understandable.

donn-xx commented:
> Please click on **Files changed** ...
> In this PR, I have already modified the description a loooong while ago, (in an attempt) to be more understandable.

Magic. I didn't know that trick. Thanks. Your description is very good!

donn-xx commented:
PS - There is some churn on the examples for `bsearch`. Here's a link:

https://github.com/godotengine/godot/pull/89280#discussion_r1517763110

Mickeon commented:
I was there 😉 thank you for the reminder though!

Mickeon commented:
How's this for an example @AThousandShips @donn-xx ?
```GDScript
func sort_by_amount(a, b):
	if a[1] < b[1]:
		return true
	return false

func _ready():
	var my_items = [["Tomato", 2], ["Kiwi", 5], ["Rice", 9]]
	
	var apple = ["Apple", 5]
	# "Apple" is inserted before "Kiwi".
	my_items.insert(my_items.bsearch_custom(apple, sort_by_amount, true), apple)
	
	var banana = ["Banana", 5]
	# "Banana" is inserted after "Kiwi".
	my_items.insert(my_items.bsearch_custom(banana, sort_by_amount, false), banana)
	
	# Prints [["Tomato", 2], ["Apple", 5], ["Kiwi", 5], ["Banana", 5], ["Rice", 9]]
	print(my_items)
```

AThousandShips commented:
I think it looks good, not convinced it's needed though, but I'd do:
```gdscript
func sort_by_amount(a, b):
    return a[1] < b[1]
```

Mickeon commented:
I chose to be explicit as some users do not "wrap around the idea" of comparisons returning `bool` (It's also consistent with the other example). I guess I could make an exception here? Binary search is not of every-day use.

donn-xx commented:
> How's this for an example @AThousandShips @donn-xx ?

Ah, I see. You pass an array to bsearch_custom. I think the answer is to make it clear that both `a` *and* `b` must be the same types (even though they are accepted as variants).

AThousandShips commented:
> both a and b must be the same types (even though they are accepted as variants).

That's not the case, you can search for anything as long as you handle it correctly

Mickeon commented:
For the love of the almighty this PR is so old the screenshot taken as a preview doesn't even match how the built-in documentation looks like anymore

donn-xx commented:
> > both a and b must be the same types (even though they are accepted as variants).
> That's not the case, you can search for anything as long as you handle it correctly

True. I was speaking about practical use. Yeah, I didn't think about probing the argument types within the func and then changing behaviour dynamically.

> In fact I think the one context where it would be good to add an example is to handle that detail, like searching for `5` and not an array element

Indeed! This would help.

Mickeon commented:
Painfully rebased after https://github.com/godotengine/godot/pull/79075 

@dalexeev see if you're comfortable with the changes, too. Look for `get_typed_class_name` `get_typed_script`, and the typed Array constructor.

dalexeev commented on doc/classes/Array.xml:
```diff
@@ -58,29 +48,32 @@
-				    var a: Array[int] = []
-				    some_func(a)
+				var numbers: Array[float] = []
+				var children: Array[Node] = get_children()
```
This is not a very good example, since here the array is not created, but an already existing one is assigned. I would suggest the following:

```suggestion
				var children: Array[Node] = [$Node1, $Node2, $Node3]
```

Mickeon replied:
Will do, with name changes to also hint towards the objects inheriting from "Node" that matters here.

dalexeev commented on doc/classes/Array.xml:
```diff
@@ -341,196 +391,188 @@
 			<param index="0" name="what" type="Variant" />
 			<param index="1" name="from" type="int" default="0" />
 			<description>
-				Searches the array for a value and returns its index or [code]-1[/code] if not found. Optionally, the initial search index can be passed.
+				Returns the index of the [b]first[/b] occurrence of [param what] in this array, or [code]-1[/code] if there are none. The search's start can be specified with [param from], continuing to the end of the array.
+				[b]Note:[/b] If you just want to know whether the array contains [param what], use [method has] ([code]Contains[/code] in C#). In GDScript, you may also use the [code]in[/code] operator.
+				[b]Note:[/b] For performance reasons, the search is affected by [param what]'s [enum Variant.Type]. For example, [code]7[/code] ([int]) and [code]7.0[/code] ([float]) are not considered equal for this method.
 			</description>
 		</method>
 		<method name="front" qualifiers="const">
 			<return type="Variant" />
 			<description>
-				Returns the first element of the array. Prints an error and returns [code]null[/code] if the array is empty.
-				[b]Note:[/b] Calling this function is not the same as writing [code]array[0][/code]. If the array is empty, accessing by index will pause project execution when running from the editor.
+				Returns the first element of the array. If the array is empty, fails and returns [code]null[/code]. See also [method back].
+				[b]Note:[/b] Unlike with the [code][][/code] operator ([code]array[0][/code]), an error is generated without stopping project execution.
 			</description>
 		</method>
 		<method name="get_typed_builtin" qualifiers="const">
 			<return type="int" />
 			<description>
-				Returns the built-in type of the typed array as a [enum Variant.Type] constant. If the array is not typed, returns [constant TYPE_NIL].
+				Returns the built-in type of the typed array as a [enum Variant.Type] constant. If the array is not typed, returns [constant TYPE_NIL]. See also [method is_typed].
 			</description>
 		</method>
 		<method name="get_typed_class_name" qualifiers="const">
 			<return type="StringName" />
 			<description>
-				Returns the [b]native[/b] class name of the typed array if the built-in type is [constant TYPE_OBJECT]. Otherwise, this method returns an empty string.
+				Returns the base class name of the typed array, if the built-in [Variant] type [constant TYPE_OBJECT]. Otherwise, returns an empty [StringName]. See also [method is_typed] and [method Object.get_class].
```
This loses the emphasis that the method only returns the built-in/native class, not a global script class. This is especially important given that the method is called "get_typed_**class_name**".

Mickeon replied:
I'll stick to calling it "built-in class name" then. The intention is to use consistent wording when referring to essentially "What `Object.get_class` returns".

mhilbrunner reviewed (approved):
I spent the better part of an hour poring over the changes. There are some small parts I could quibble about, but with a change this size, that is always going to be the case. None IMO are important enough to block this further, seeing how long it has been open.

What matters is that overall, this is a large improvement IMO, and none of the few things above matter too much. Overall, this is more clear, more carefully worded and links relevant info. My take is to get this in. :)

akien-mga commented:
Thanks!

Mickeon commented:
I have been freed from the curse.
